File: from-0801-to-0900/0881-boats-to-save-people/solution.go
func numRescueBoats(people []int, limit int) int {
    result := 0
    sort.Ints(people)
    left, right := 0, len(people) - 1
    for left <= right {
        if left == right {
            return result + 1
        }
        
        if people[right] + people[right - 1] <= limit {
            right -= 2
        } else if people[right] + people[left] <= limit {
            right -= 1
            left += 1
        } else {
            right -= 1
        }
        result += 1
    }
    return result
}

File: from-0801-to-0900/0881-boats-to-save-people/solution.py
class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        peps = sorted(people)
        ans = 0
        ph = 0
        pt = len(peps)-1
        
        while True:
            if ph == pt:
                return ans + 1
            if ph > pt:
                return ans
            group_weight = peps[pt]
            if peps[pt] + peps[pt-1] <= limit:
                ans += 1
                pt -= 2
            elif peps[pt] + peps[ph] <= limit:
                ans += 1
                pt -= 1
                ph += 1
            else:
                ans += 1
                pt -= 1